File: algo/src/test/java/com/hartwig/actin/algo/evaluation/molecular/HasSpecificHLATypeTest.java
package com.hartwig.actin.algo.evaluation.molecular;

import static com.hartwig.actin.algo.evaluation.EvaluationAssert.assertMolecularEvaluation;

import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.molecular.datamodel.immunology.HlaAllele;
import com.hartwig.actin.molecular.datamodel.immunology.ImmutableHlaAllele;
import com.hartwig.actin.molecular.datamodel.immunology.ImmutableMolecularImmunology;
import com.hartwig.actin.molecular.datamodel.immunology.MolecularImmunology;

import org.jetbrains.annotations.NotNull;
import org.junit.Test;

//TODO: Fix test
public class HasSpecificHLATypeTest {

    @Test
    public void canEvaluate() {
        String correct = "A*02:01";
        HasSpecificHLAType function = new HasSpecificHLAType(correct);

        assertMolecularEvaluation(EvaluationResult.UNDETERMINED,
                function.evaluate(MolecularTestFactory.withMolecularImmunology(create(false))));
        assertMolecularEvaluation(EvaluationResult.UNDETERMINED,
                function.evaluate(MolecularTestFactory.withMolecularImmunology(create(false, correct))));

        assertMolecularEvaluation(EvaluationResult.WARN,
                function.evaluate(MolecularTestFactory.withMolecularImmunology(create(true, correct))));

        assertMolecularEvaluation(EvaluationResult.FAIL,
                function.evaluate(MolecularTestFactory.withMolecularImmunology(create(true, "other"))));
    }

    @NotNull
    private static MolecularImmunology create(boolean isReliable, @NotNull String allele) {
        HlaAllele hlaAllele = ImmutableHlaAllele.builder().name(allele).tumorCopyNumber(0D).hasSomaticMutations(false).build();
        return create(isReliable, hlaAllele);
    }

    @NotNull
    private static MolecularImmunology create(boolean isReliable, @NotNull HlaAllele... alleles) {
        return ImmutableMolecularImmunology.builder().from(create(isReliable)).addHlaAlleles(alleles).build();
    }

    @NotNull
    private static MolecularImmunology create(boolean isReliable) {
        return ImmutableMolecularImmunology.builder().isReliable(isReliable).build();
    }
}

File: algo/src/test/kotlin/com/hartwig/actin/algo/evaluation/molecular/HasSpecificHLATypeTest.kt
package com.hartwig.actin.algo.evaluation.molecular

import com.hartwig.actin.algo.evaluation.EvaluationAssert.assertMolecularEvaluation
import com.hartwig.actin.datamodel.algo.EvaluationResult
import com.hartwig.actin.datamodel.molecular.orange.immunology.HlaAllele
import org.junit.Test

private val CORRECT_HLA = HlaAllele(name = "A*02:01", tumorCopyNumber = 1.0, hasSomaticMutations = false)

class HasSpecificHLATypeTest {

    private val function = HasSpecificHLAType(CORRECT_HLA.name)

    @Test
    fun `Should pass if correct HLA allele present`() {
        assertMolecularEvaluation(EvaluationResult.PASS, function.evaluate(MolecularTestFactory.withHlaAllele(CORRECT_HLA)))
    }

    @Test
    fun `Should evaluate to undetermined if immunology results are unreliable`() {
        assertMolecularEvaluation(
            EvaluationResult.UNDETERMINED,
            function.evaluate(MolecularTestFactory.withUnreliableMolecularImmunology())
        )
    }

    @Test
    fun `Should evaluate to undetermined if no WGS results present`() {
        assertMolecularEvaluation(EvaluationResult.UNDETERMINED, function.evaluate(MolecularTestFactory.withIHCTests()))
    }

    @Test
    fun `Should evaluate to undetermined if correct HLA allele present but record does not have sufficient quality`() {
        assertMolecularEvaluation(
            EvaluationResult.UNDETERMINED,
            function.evaluate(MolecularTestFactory.withHlaAlleleAndInsufficientQuality(CORRECT_HLA))
        )
    }

    @Test
    fun `Should warn if correct HLA allele present but tumor copy number is less than 0,5`() {
        assertMolecularEvaluation(
            EvaluationResult.WARN,
            function.evaluate(MolecularTestFactory.withHlaAllele(CORRECT_HLA.copy(tumorCopyNumber = 0.0)))
        )
    }

    @Test
    fun `Should warn if correct HLA allele present but also somatic mutations present`() {
        assertMolecularEvaluation(
            EvaluationResult.WARN,
            function.evaluate(MolecularTestFactory.withHlaAllele(CORRECT_HLA.copy(hasSomaticMutations = true)))
        )
    }

    @Test
    fun `Should fail if correct HLA allele not present`() {
        assertMolecularEvaluation(
            EvaluationResult.FAIL,
            function.evaluate(MolecularTestFactory.withHlaAllele(CORRECT_HLA.copy(name = "other")))
        )
        assertMolecularEvaluation(
            EvaluationResult.FAIL,
            function.evaluate(MolecularTestFactory.withHlaAlleleAndInsufficientQuality(CORRECT_HLA.copy(name = "other")))
        )
    }
}